### ISSUE TITLE
Diagnostics can be added after they are written

### ISSUE THREAD

**Issue by mikeminutillo** [Particular/NServiceBus#6656]
NServiceBus provides an api `AddStartupDiagnosticsSection(..)` for adding to the diagnostic file written at startup. This can be called after the diagnostics have already been written (for example [in the transport initialization](https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/issues/744)). Sections added after the diagnostics have already been written are only ever kept in memory, and are effectively lost.

It would be better if the diagnostics section were frozen after it is written. Subsequent attempts to add new sections could throw an exception so we could detect this situation earlier.

**Comment by danielmarbach** [comment]
@mikeminutillo as far as I understand, the problem is even deeper than just preventing to add diagnostics. Even though the transport gets passed the `hostSettings` which contains the `StartupDiagnostics` writing capability, it can never add anything there because the diagnostics have already been written, which means effectively a transport cannot add any diagnostics. The only sane way to add diagnostics entry would be in

```
        public override async Task<TransportInfrastructure> Initialize(HostSettings hostSettings,
            ReceiveSettings[] receivers, string[] sendingAddresses, CancellationToken cancellationToken = default)
```

Doesn't that mean the Core order requires changes? // @SzymonPobiega

**Comment by SzymonPobiega** [comment]
> Doesn't that mean the Core order requires changes?

I am afraid so.

**Comment by mikeminutillo** [comment]
@danielmarbach Yes. There is currently no way for a transport to add diagnostic entries that will be written. I believe that ASB is the only transport that does.

If we froze the diagnostic entries and threw we might have discovered this issue sooner.

**Comment by PhilBastian** [comment]
note that the example specified (transport initialisation being called after diagnostics are written) is [no longer an issue](https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/issues/744#issuecomment-3295325923). The core issue reported here, that `AddStartupDiagnosticsSection` can be called after the diagnostics have been written, still remains
